File: AIDA64Remote-iOS/AIDA64Remote/Data/SensorParser.swift
import Foundation

enum SensorParser {
    private static let pageMarker = try! NSRegularExpression(pattern: #"^Page\d+$"#, options: .caseInsensitive)

    static func parse(_ dataLine: String) -> [SensorItem] {
        var payload = dataLine
        if payload.hasPrefix("data:") {
            payload = String(payload.dropFirst(5))
        }
        payload = payload.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !payload.isEmpty else { return [] }

        return payload
            .components(separatedBy: "{|}")
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .compactMap { segment -> SensorItem? in
                guard let separator = segment.firstIndex(of: "|") else { return nil }
                let id = String(segment[..<separator]).trimmingCharacters(in: .whitespacesAndNewlines)
                guard !id.isEmpty, !isPageMarker(id) else { return nil }
                let rawValue = String(segment[segment.index(after: separator)...])
                let value = rawValue
                    .split(separator: "|", maxSplits: 1, omittingEmptySubsequences: false)
                    .first
                    .map(String.init)?
                    .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
                return SensorItem(id: id, value: value)
            }
    }

    static func toMap(_ items: [SensorItem]) -> [String: String] {
        Dictionary(uniqueKeysWithValues: items.map { ($0.id, $0.value) })
    }

    private static func isPageMarker(_ id: String) -> Bool {
        let range = NSRange(id.startIndex..<id.endIndex, in: id)
        return pageMarker.firstMatch(in: id, options: [], range: range) != nil
    }
}

File: AIDA64Remote-Android/app/src/main/java/com/example/aida64remote/data/SensorParser.kt
package com.example.aida64remote.data

import com.example.aida64remote.model.SensorItem

object SensorParser {
    private val pageMarker = Regex("^Page\\d+$", RegexOption.IGNORE_CASE)

    fun parse(dataLine: String): List<SensorItem> {
        val payload = dataLine
            .removePrefix("data:")
            .trim()
            .ifEmpty { return emptyList() }

        return payload
            .split("{|}")
            .asSequence()
            .map { it.trim() }
            .filter { it.isNotEmpty() }
            .mapNotNull { segment ->
                val separator = segment.indexOf('|')
                if (separator <= 0) return@mapNotNull null
                val id = segment.substring(0, separator).trim()
                if (id.isEmpty() || pageMarker.matches(id)) return@mapNotNull null
                val rawValue = segment.substring(separator + 1)
                val value = rawValue.substringBefore('|').trim()
                SensorItem(id = id, value = value)
            }
            .toList()
    }

    fun toMap(items: List<SensorItem>): Map<String, String> =
        items.associate { it.id to it.value }
}
